### ISSUE TITLE
head pose estimation, aflw dataset

### ISSUE THREAD

**Issue by NieHw** [mpatacchiola/deepgaze#60]
Thanks for your work. I have a question about the CNN_Head pose estimation. Before I use the AFLW dataset to train the network, do I need to resize the face picture to 64*64? How to complete it?

**Comment by mpatacchiola** [comment]
Hi @NieHw 

Yes in general it is a good practice to crop, resize, and scale the input images.

The AFLW dataset has associated bounding boxes that define the position of each face. Using these coordinates you can crop the face. You have to be careful because some faces are close to the border of the picture and in this case to get a correct crop you have to translate the box in a valid area.

To resize the cropped subframe to 64x64 pixels you can use the OpenCV method called `resize`.

**Comment by NieHw** [comment]
Thanks for your reply. I will google it. And I want to implement the experiment of your paper as practice. I have some questions about the cross-validation when you used the Prima Head Pose Dataset. Can you please give me your training code? Thank you again.

**Comment by NieHw** [comment]
Hi @mpatacchiola 
 I got some information in .sqlite file about the bounding boxes, "x, y, h, w" these four coordinates. But I noticed that there were some negative coordinates numbers. And when my crop programer these pictures(negative coordinates), it will creat a damaged picture. Do u know how to avoid it? Thank you very much!

**Comment by NieHw** [comment]
Hi @mpatacchiola
I got some information in .sqlite file about the bounding boxes, "x, y, h, w" these four coordinates. But I noticed that there were some negative coordinates numbers. And when my crop program process these pictures(with negative coordinates), the program will create a damaged picture. Do u know how to avoid it? Thank you very much!

**Comment by NieHw** [comment]
Hi @mpatacchiola
I got some information in .sqlite file about the bounding boxes, "x, y, h, w" these four coordinates. But I noticed that there were some negative coordinates numbers. And when my crop program processed these pictures(with negative coordinates), the program would create a damaged picture. Do u know how to avoid it? Thank you very much!

**Comment by NieHw** [comment]
Hi @mpatacchiola 
I used the ‘ex_prima_parser.py' to handle the Prima Head Pose Dataset, and I noticed that you separated the tilt_label and pan_label. When I used tilt_label to train my cnn, after tf.nn.softmax(), it will putout [1,1,1,1....] or [0,0,0,0,....] as predicition. I don't understand it. How do I to train the network?
Help me please, thank you very much. Sorry to bother you.

**Comment by mpatacchiola** [comment]
Hi @NieHw you have to think about head pose estimation not like a classification problem but like a regression problem. You have to use a Sigmoid or Tanh activation funciton in the output layer and use the Squared error loss function to compare your output and the target.

However, the Prima dataset has discrete poses and you can think of dealing with it as a classification problem. In this case you should have an output unit for each class and use a Softmax activation function in the output, and the crossentropy as loss.

**Comment by NieHw** [comment]
Thanks for your patient reply. I just used regression (not softmax) as a output and I get a result close to yours in paper.

**Comment by speculaas** [comment]
Dear NieHw, 
Thanks for your help!
Can you share which script you used?
I tried ex_dnn_head_pose_estimation_training.py, but after 5 million steps, 
I use the result, dnn_1600i_4h_3o-5000000, to estimate different faces of different poses.

The inputs' poses varies, but the resulting ckpt's estimation always the same. 

mpatacchiola said I might stuck at a local minimum. 
But currently I am not sure what to do with it. 

So hope you can share how you achieve similar result to the paper. 

Another point, the paper uses cnn to do regression, but ex_dnn_head_pose_estimation_training 
uses fully connected layers. 
My doubt about this is whether fully connected layers can learn angles?
I think to learning angles, the input to the network better to 2D?

BR,
Jimmy

![deepgaze_training_result 2018 09 06 r 03 20 pm](https://user-images.githubusercontent.com/31201176/45141761-5d8fc680-b1e9-11e8-9258-174692c3be90.JPG)
